quaddtype/numpy_quaddtype/src/casts.cpp

```diff
@@ -151,38 +157,407 @@ quad_to_quad_strided_loop_aligned(PyArrayMethod_Context *context, char *const da
+                                 npy_intp *view_offset)
 {
-    PyErr_SetString(PyExc_TypeError, 
-        "Void to QuadPrecision cast is not implemented");
+    PyErr_SetString(PyExc_TypeError, "Void to QuadPrecision cast is not implemented");
     return (NPY_CASTING)-1;
 }
 
 static int
 void_to_quad_strided_loop(PyArrayMethod_Context *context, char *const data[],
-                         npy_intp const dimensions[], npy_intp const strides[],
-                         void *NPY_UNUSED(auxdata))
+                          npy_intp const dimensions[], npy_intp const strides[],
+                          void *NPY_UNUSED(auxdata))
 {
     PyErr_SetString(PyExc_RuntimeError, "void_to_quad_strided_loop should not be called");
     return -1;
 }
 
+// Unicode/String to QuadDType casting
+static NPY_CASTING
+unicode_to_quad_resolve_descriptors(PyObject *NPY_UNUSED(self), PyArray_DTypeMeta *dtypes[2],
+                                    PyArray_Descr *given_descrs[2], PyArray_Descr *loop_descrs[2],
+                                    npy_intp *view_offset)
+{
+    Py_INCREF(given_descrs[0]);
+    loop_descrs[0] = given_descrs[0];
+
+    if (given_descrs[1] == NULL) {
+        loop_descrs[1] = (PyArray_Descr *)new_quaddtype_instance(BACKEND_SLEEF);
+        if (loop_descrs[1] == nullptr) {
+            return (NPY_CASTING)-1;
+        }
+    }
+    else {
+        Py_INCREF(given_descrs[1]);
+        loop_descrs[1] = given_descrs[1];
+    }
+
+    return NPY_UNSAFE_CASTING;
+}
+
+static int
+unicode_to_quad_strided_loop_unaligned(PyArrayMethod_Context *context, char *const data[],
+                                       npy_intp const dimensions[], npy_intp const strides[],
+                                       void *NPY_UNUSED(auxdata))
+{
+    npy_intp N = dimensions[0];
+    char *in_ptr = data[0];
+    char *out_ptr = data[1];
+    npy_intp in_stride = strides[0];
+    npy_intp out_stride = strides[1];
+
+    PyArray_Descr *const *descrs = context->descriptors;
+    QuadPrecDTypeObject *descr_out = (QuadPrecDTypeObject *)descrs[1];
+    QuadBackendType backend = descr_out->backend;
+
+    // Unicode strings are stored as UCS4 (4 bytes per character)
+    npy_intp unicode_size_chars = descrs[0]->elsize / 4;
+
+    while (N--) {
+        // Temporary buffer to convert UCS4 to null-terminated char string
+        char temp_str[QUAD_STR_WIDTH + 1];
+        npy_intp copy_len =
+                unicode_size_chars < QUAD_STR_WIDTH ? unicode_size_chars : QUAD_STR_WIDTH;
+        // Convert UCS4 characters to ASCII/char
+        Py_UCS4 *ucs4_str = (Py_UCS4 *)in_ptr;
+        npy_intp i;
+        for (i = 0; i < copy_len; i++) {
+            Py_UCS4 c = ucs4_str[i];
+
+            // reject non-ASCII characters
+            if (c > 127) {
+                PyErr_Format(PyExc_ValueError,
+                             "Cannot cast non-ASCII character '%c' to QuadPrecision", c);
+                return -1;
+            }
```

Comment:
won't the string-conversion functions called by `cstring_to_quad` catch this?

By the way, you might take inspiration from the `string_to_long_double` function in numpy:

https://github.com/numpy/numpy/blob/fabf1844667c81344ec0d0ab4958cde9cb9a845c/numpy/_core/src/multiarray/arraytypes.c.src#L533-L595

You may also want to simply copy/paste and vendor the code in `NumPyOS_ascii_strtold` into this project to get a parser that act exactly like numpy's.

Reply:
Hmm reading the code, there is not much difference, we directly working with UCS but NumPy converts it first to UTF-8 ( we can do this) and special case handling, white spaces, etc are already being handled by the `Sleef_strotoq` and `strtold` except the handling of `locale` which we might not be able to do as 
- SLEEF's conversion function does no support that
- `strtold_l` (used by `NumPyOS_ascii_strtold`)  is not C standard (a posix extension)

Reply:
https://github.com/numpy/numpy/blob/fabf1844667c81344ec0d0ab4958cde9cb9a845c/numpy/_core/src/multiarray/arraytypes.c.src#L579

this line is the issue